synapse/handlers/presence.py

```diff
@@ -1430,37 +1430,88 @@ async def _handle_state_delta(self, room_id: str, deltas: List[JsonDict]) -> Non
         #   2. presence states of newly joined users to all remote servers in
         #      the room.
         #
-        # TODO: Only send presence states to remote hosts that don't already
+        # However, only send presence states to remote hosts that don't already
         # have them (because they already share rooms).
 
         # Get all the users who were already in the room, by fetching the
         # current users in the room and removing the newly joined users.
         users = await self.store.get_users_in_room(room_id)
         prev_users = set(users) - newly_joined_users
 
+        async def is_only_one_room_shared(
+            current_user_id: str, other_user_id: str
+        ) -> bool:
+            """Check to see if there is more than one shared room between these users"""
+            pair_of_users_to_check = frozenset((current_user_id, other_user_id))
+            # By learning if these two users share any other room, we can tell
+            # if user_id's server needs to receive presence or not
+            rooms_shared = await self.store.get_mutual_rooms_between_users(
+                pair_of_users_to_check
+            )
+
+            # They always share at least one room, the current one. As such this will
+            # never be zero
+            if len(rooms_shared) < 2:
+                return True
+
+            return False
+
         # Construct sets for all the local users and remote hosts that were
         # already in the room
         prev_local_users = []
-        prev_remote_hosts = set()
+        prev_remote_hosts: Set[str] = set()
+
         for user_id in prev_users:
             if self.is_mine_id(user_id):
                 prev_local_users.append(user_id)
             else:
-                prev_remote_hosts.add(get_domain_from_id(user_id))
+                # Only save a host to send presence to if this user shares no other
+                # rooms with other users this server knows about
+                host = get_domain_from_id(user_id)
+
+                # Remove current user_id from the set to check against
+                prev_users_excluding_current_user = prev_users.copy() - {user_id}
+
+                for other_user_id in prev_users_excluding_current_user:
+                    # If this host has been seen in a list already, skip it.
+                    if host not in prev_remote_hosts:
+                        if await is_only_one_room_shared(user_id, other_user_id):
```

Comment:
> I'm not following this, why are we comparing all users previously in the room to see if they share a room?

Excellent question. We want to build a full list of all remote servers in the room, not the actual users. (We shortcut out so not to hit the database more than necessary). *But*, we want the list to excluded some servers, hence the comparison. The list needs to be both as full as possible and as empty as possible at the same time.

> Notably, I think both `user_id` and `other_user_id` can be remote users?

Not just can be, they are. It means that we know the host from `user_id` is in this room, *but* if `other_user_id`'s host(which we know is in this room obviously) is in *yet another* room *that we know about* because it's federated to this server, than `user_id`'s host can be excluded, since they are already receiving presence. For example:
`@me:myserver`, `@alice:server1` and `@bob:server2` are already in `Synapse Admins`.
`@me:myserver`, `@bob:server2` and `@charlie:server2` are in `Matrix HQ`.
`@alice:server1` is joining `Matrix HQ`.
Because we know `server1` is already receiving presence info about `@bob:server2`, it can be excluded from the list of `prev_remote_hosts`.

In retrospect, perhaps an additional list maintaining all the hosts we know can be excluded would also be a good idea, to cut down on database hits? However, I see a potential gotcha here, as what happens to `@charlie:server2`? If we maintain an excluded list as well, he would never be checked against and therefore(at least at first) not receive the new presence data.(Depending of course on if `@bob` or `@charlie` is checked first)

> I'd have thought that at least one iteration here should be across the newly joined users or hosts?

Newly joined(either local or remote) is dealt with in the next block below, starting at about line 1496. 

It could actually be argued that presence would be retrieved on the first(next?) sync that occurs after the room is fully joined anyways, so why are we bothering with this at all?

Reply:
I had an epiphany while at work tonight; the `prev_remote_hosts` only needs to be built if one of the newly joined users is local. If that's what you were getting at with your question, mission accomplished 😉. I'll do a refactor in a bit.

Reply:
> I had an epiphany while at work tonight; the `prev_remote_hosts` only needs to be built if one of the newly joined users is local. If that's what you were getting at with your question, mission accomplished 😉. I'll do a refactor in a bit.

I still want to make this optimization a thing, but perhaps it's best left to another PR. If that's ok?